### PR TITLE
Adding RSC wrapper for InputStream

### PR DESCRIPTION
**Pull request by vgkholla**
Adding an implementation of `ReadableStreamChannel` that wraps an `InputStream`. `InputStream` implementations are blocking and widely used and this, along with `ReadableChannelInputStream`, will help those using the `Router` or rest framework to easily interconvert to and from `InputStream`.

Built and formatted.

**Primary reviewer(s): Either of @sameerapadhye, @saravindpsg, @fhlawaczek or @ggronkowski**
**Secondary reviewer: @pnarayanan**
**Expected time to review: 40-60 mins**

**Unit testing**

Class | Class, % | Method, % | Line, %
------- | ------------ | -------------- | ----------
`InputStreamReadableStreamChannel` | 100% (4/ 4) | 100% (15/ 15) | 100% (63/ 63)
`ReadableStreamChannelInputStream` | 100% (2/ 2) | 100% (9/ 9) | 94.7% (54/ 57)

### CODE REVIEW

File: ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+  private int fillBuffer()
+      throws IOException {
+    int totalRead = 0;
+    int currentRead;
+    do {
+      currentRead = inputStream.read(buffer, totalRead, buffer.length - totalRead);
+      if (currentRead > 0) {
+        totalRead += currentRead;
+      }
+    } while (currentRead < buffer.length && currentRead > 0);
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java]: Thats right. This should be `totalRead`. Messed it up when changing variable names. Thanks for catching !

File: ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                if (exception != null) {
+                  callbackWrapper.invokeCallback(exception);
+                } else {
+                  callbackWrapper.updateBytesRead(result);
+                  writeToChannel(writableChannel);
+                }
+              }
+            });
+          } else {
+            // the input stream has been emptied.
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java]: `InputStream` read methods can [never](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[],%20int,%20int)) return 0.

File: ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java (hunk 1 above)
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java]: Also, since `InputStream` never returns 0, `currentRead > 0` is the same as `currentRead !=1`. But I will change it because the latter is probably clearer.

File: ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java (hunk 2 above)
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java]: Just one caveat - it returns 0 if the length asked to be read is 0.

**Comment by vgkholla** [comment]
@pnarayanan Addressed comments

**Comment by pnarayanan** [comment]
Will take a second look when I get a chance, but any of the primary reviewers planning to look at this?

**Comment by saravindpsg** [comment]
I am looking into this, thanks @pnarayanan

**Comment by saravindpsg** [comment]
:shipit:

File: ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java [hunk 3]
```diff
@@ -127,6 +127,16 @@ private boolean loadData()
   }
 
   /**
+   * Keeps track of the bytes read.
+   * @param count the number of bytes read in this read.
+   */
+  private void reportBytesRead(int count) {
+    if (bytesAvailable - count >= 0) {
```
Comment (pnarayanan) [inline ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java]: `if (bytesAvailable >=  count)` looks simpler?

File: ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java [hunk 4]
```diff
@@ -40,7 +40,7 @@
    */
```
Comment (pnarayanan) [inline ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java]: just curious: in the comment for this class, why do you say "possibly non-blocking `ReadableStreamChannel`"? A `ReadableStreamChannel` is never blocking, so I think you mean something else here - you might want to clarify.

File: ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java [hunk 5]
```diff
@@ -55,7 +55,7 @@ public int read()
     int data = -1;
```
Comment (pnarayanan) [inline ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java]: `available()` should not return -1 (which I think it can in this implementation). Also, whatever `available()` returns should be readable without blocking - so returning all of `bytesAvailable` (or Integer.MAX) would not be the right thing as `loadData()` can block in trying to read that much.

File: config/PartitionLayout.json [hunk 6]
```diff
@@ -10,7 +10,7 @@
                 {
                     "hostname": "localhost",
                     "mountPath": "/tmp",
-                    "port": 6667
+                    "port": 15088
```
Comment (pnarayanan) [inline config/PartitionLayout.json]: why?

File: config/HardwareLayout.json [hunk 7]
```diff
@@ -14,10 +14,10 @@
                     ],
                     "hardwareState": "AVAILABLE",
                     "hostname": "localhost",
-                    "port": 6667
+                    "port": 15088
                 }
             ],
-            "name": "Datacenter"
+            "name": "Data-Center"
```
Comment (pnarayanan) [inline config/HardwareLayout.json]: interesting that you are making this change too - I have discarded this change in the other PR, but yes let us make this change as we have certain other hardware layout files that have this change.

File: config/HardwareLayout.json [hunk 8]
```diff
@@ -14,10 +14,10 @@
                     ],
                     "hardwareState": "AVAILABLE",
                     "hostname": "localhost",
-                    "port": 6667
+                    "port": 15088
```
Comment (pnarayanan) [inline config/HardwareLayout.json]: why make this change?

File: ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java (hunk 4 above)
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java]: Removed that qualifier.

File: ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java (hunk 5 above)
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java]: `available()` cannot return -1 since the constructor initializes it correctly and `reportBytesRead()` makes sure it doesn't go under 0. 

Your other point is correct and I've now modified the function to return whatever is remaining in the current chunk.

File: config/HardwareLayout.json (hunk 7 above)
Comment (vgkholla) [inline config/HardwareLayout.json]: Will make these changes in another PR and roll these back for now.

File: ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java [hunk 9]
```diff
@@ -107,7 +107,14 @@ private boolean loadData()
         asyncWritableChannel.resolveOldestChunk(null);
       }
       try {
-        currentChunk = asyncWritableChannel.getNextChunk();
+        while (true) {
```
Comment (pnarayanan) [inline ambry-commons/src/main/java/com.github.ambry.commons/ReadableStreamChannelInputStream.java]: I think this method should throw the `InterruptedException` and let the caller deal with it.
Also, are you sure this method cannot be made simpler? I find certain things repetitive in here.

**Comment by pnarayanan** [comment]
Would be great if you could refactor `loadData()` to make it a bit more simple/intuitive. Other than that LGTM.
